feat(namespaces): Extend namespace grpc update handler for jsonschema updates

osalama7 opened the pull request:
- Handle Del and Put ops explicitly
- Add new helper method for json schema validation in json schema wrapper
- Improve Unit tests coverage for json schema and wrapper
- Fix tests for namespace Dao and test Add and update changes
- Ensure publishing broker events on Update

Testing the update endpoint:
```
postman request PUT 'https://localhost:8080/a/user-meta/namespace' \
  --header 'Content-Type: application/json' \
  --header 'Authorization: Bearer <token>' \
  --body '{
  "Operation": "PUT",
  "Namespaces": [
    {
      "Indexable": true,
      "JsonDefinition": "{\"type\":\"string\"}",
      "JsonSchema": {
        "$id": "",
        "additionalProperties": false,
        "properties": {
          "text": {
            "maxLength": 4,
            "minLength": 1,
            "type": "string"
          }
        },
        "title": "Text",
        "type": "object",
        "required": ["text"]
      },
      "Label": "Text",
      "Namespace": "usermeta-text"
    }
  ]
}' \
  --auth-bearer-token '<token>'

osalama7 committed:
feat(namespaces): Extend namespace grpc handler
- Handle Del and Put ops explicitly
- Add new helper method for json schema validation in json schema wrapper
- Improve Unit tests coverage for json schema and wrapper
- Fix tests for namespace Dao and test Add and update changes

- Add a todo to handle Validation in another layer.

fix(namespace) re order import

clean(namespace): improve comments

fix(namespace): Improve error returns

- Remove unused pbstruct return type from GetJsonSchema
- Fix Unit tests.

osalama7 commented on idm/meta/json_schema/schema.go:
```diff
@@ -41,3 +43,65 @@ func InferBytes[T any](opts *ForOptions) ([]byte, error) {
+		return err
+	}
+
+	isObjectLike := false
+	if t, ok := m["type"]; ok {
+		if ts, ok := t.(string); ok && ts == "object" {
+			isObjectLike = true
+		}
+	}
+	if _, ok := m["required"]; ok {
+		isObjectLike = true
+	}
+	if _, ok := m["additionalProperties"]; ok {
+		isObjectLike = true
+	}
+
+	// If schema appears to describe an object, ensure 'properties' exists and is non-empty.
+	if isObjectLike {
+		props, ok := m["properties"]
+		if !ok {
+			return nil
+		}
+		if pm, ok := props.(map[string]any); !ok || len(pm) == 0 {
+			return nil
+		}
+	}
+
+	return nil
```
Basic validation of JsonSchema structure, Making sure we don't store invalid JsonSchemas.
I could think of a data driven way of validating at this point of schema update, i.e. using existing tags. but it wont be very stable if JsonSchema type change is not resilient.

osalama7 commented on idm/meta/dao/sql/ns-sql.go:
```diff
@@ -96,7 +95,12 @@ func (u *MetaNamespace) From(res *idm.UserMetaNamespace) *MetaNamespace {
 
-func (u *MetaNamespace) FromExisting(res *idm.UserMetaNamespace) *MetaNamespace {
+func (u *MetaNamespace) FromExisting(res *idm.UserMetaNamespace) (*MetaNamespace, error) {
+	jssc := json_schema.ValidateSchemaFromPbStruct(res.JsonSchema)
```
This validation could be moved to another layer.

cdujeu replied:
not a big issue to me, better late than ignore errors ;-)

cdujeu commented on idm/meta/dao/sql/ns-sql.go:
```diff
@@ -154,9 +158,28 @@ func (s *nsSqlImpl) Migrate(ctx context.Context) error {
+		return nil
+	}
+	// Insert
+	tx1 := s.Session(ctx).Clauses(clause.OnConflict{UpdateAll: true}).Create((&MetaNamespace{}).From(ns))
```
shall we keep the OnConflict here as we are now sure it's a CREATE?

cdujeu commented on idm/meta/grpc/handler.go:
```diff
@@ -315,18 +315,25 @@ func (h *Handler) UpdateUserMetaNamespace(ctx context.Context, request *idm.Upda
 	}
 	if request.Operation == idm.UpdateUserMetaNamespaceRequest_PUT {
 		for _, metaNameSpace := range request.Namespaces {
+			broker.MustPublish(ctx, common.TopicIdmEvent, &idm.ChangeEvent{
```
Why publishing a DELETE here?

cdujeu commented on idm/meta/dao/sql/ns-sql.go:
```diff
@@ -154,9 +158,28 @@ func (s *nsSqlImpl) Migrate(ctx context.Context) error {
-		return nsTag(tx.Error)
+	// Update existing
+	var ex *MetaNamespace
+	tx0 := s.Session(ctx).Where(&MetaNamespace{Namespace: string(ns.Namespace)}).First(&ex)
```
naive question : why do we need string() conversion here?

osalama7 replied:
Left over from a debugging session, removed

osalama7 committed:
fix(namespace): Refactor dao method Add
- Change the function name to Upsert
- Fix tests
- Return bool for update vs create and catch in grpc handler
- Ensure Broker publishes Update and Create based on Dao method return
